source/common/tcp_proxy/upstream.cc

```diff
@@ -74,13 +93,16 @@ TcpUpstream::onDownstreamEvent(Network::ConnectionEvent event) {
   return nullptr;
 }
 
-HttpUpstream::HttpUpstream(Tcp::ConnectionPool::UpstreamCallbacks& callbacks,
+HttpUpstream::HttpUpstream(HttpConnPool& http_conn_pool,
+                           Http::StreamDecoderFilterCallbacks& decoder_callbacks,
+                           Router::Route& route, Tcp::ConnectionPool::UpstreamCallbacks& callbacks,
                            const TunnelingConfigHelper& config,
                            StreamInfo::StreamInfo& downstream_info)
-    : config_(config), downstream_info_(downstream_info), response_decoder_(*this),
+    : config_(config), downstream_info_(downstream_info), parent_(http_conn_pool),
+      decoder_filter_callbacks_(decoder_callbacks), route_(&route), response_decoder_(*this),
       upstream_callbacks_(callbacks) {}
 
-HttpUpstream::~HttpUpstream() { resetEncoder(Network::ConnectionEvent::LocalClose); }
+HttpUpstream::~HttpUpstream() = default;
```

Comment:
why this change? are we calling clean up early and is moving that safe for legacy code?

Reply:
`resetEncoder` is made a virtual function now and is being overridden by `CombinedUpstream` . clang-tidy complains if virtual function is invoked in destructor. Hence intoduced `cleanUp` and calling it from the parent's destructor.